src/symmetric_memory.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include "mococrw/symmetric_memory.h"
+
+#include <queue>
+#include <cassert>
+
+namespace mococrw
+{
+void QueueOfVectorsMemoryStrategy::write(const std::vector<uint8_t> &&chunk)
+{
+    _queue.push_back(chunk);
+    _totalBytesStored += chunk.size();
+}
+
+std::vector<uint8_t> QueueOfVectorsMemoryStrategy::read(size_t requestedBufferSize)
+{
+    if (_queue.empty()) {
+        return std::vector<uint8_t>();
+    }
+
+    if (_queue.front().size() == requestedBufferSize) {
+        return _readCompleteBlock();
+    } else if (_queue.front().size() > requestedBufferSize) {
+        return _readPartialBlock(requestedBufferSize);
+    } else {
+        std::vector<uint8_t> bufferToReturn;
+        bufferToReturn.reserve(requestedBufferSize);
+
+        // First, read as much complete blocks as we can
+        while (!_queue.empty() &&
+               (_queue.front().size() <= (requestedBufferSize - bufferToReturn.size()))) {
+            auto block = _readCompleteBlock();
+            std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
+        }
+        // Then, read partial block if needed
+        if (bufferToReturn.size() < requestedBufferSize && !_queue.empty()) {
+            auto block = _readPartialBlock(requestedBufferSize - bufferToReturn.size());
+            std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
```

Comment:
This code causes 1 unnecessary copy of the data in the "block" vector.

Reply:
Not sure if I get your point. We have to append last block of data to the result buffer. How can we do it without a copy?

Reply:
We can move the first complete block, but it looks like premature optimization to me. We would not be able to reserver() memory then. @garw  pls explain

Reply:
Done.